source/grid/manifold.cc

```diff
@@ -892,26 +894,45 @@ FlatManifold<dim, spacedim>::normal_vector(
   const unsigned int facedim = dim - 1;
 
   Point<facedim> xi;
-  for (unsigned int i = 0; i < facedim; ++i)
-    xi[i] = 1. / 2;
+
+  const bool hyper_cube =
+    face->reference_cell_type() == ReferenceCell::get_hypercube(facedim);
+
+  if (hyper_cube)
+    {
+      for (unsigned int i = 0; i < facedim; ++i)
+        xi[i] = 1. / 2;
+    }
+  else
+    {
+      for (unsigned int i = 0; i < facedim; ++i)
+        xi[i] = 1. / 3;
+    }
+
+  Simplex::FE_P<facedim> fe(1);
 
   const double        eps = 1e-12;
   Tensor<1, spacedim> grad_F[facedim];
   unsigned int        iteration = 0;
   while (true)
     {
       Point<spacedim> F;
-      for (const unsigned int v : GeometryInfo<facedim>::vertex_indices())
-        F += face->vertex(v) *
-             GeometryInfo<facedim>::d_linear_shape_function(xi, v);
+      for (const unsigned int v : face->vertex_indices())
+        F +=
+          face->vertex(v) *
+          (hyper_cube ? GeometryInfo<facedim>::d_linear_shape_function(xi, v) :
```

Comment:
I am not sure why `GeometryInfo` is used here? Isn't it the same as in `FE_Q`?

Reply:
Yes, but it is way too expensive to set up an `FE_Q` here just for the purpose of evaluating the shape functions. Note that this is an inner function that is called for many points on each face of the mesh. I think the right approach would be to add a function that can evaluate the linear basis for all reference cell types that we can query similar to `d_linear_shape_function`.

Reply:
@kronbichler What do you think about the updated version?